File: LinkedList Java and Kotlin/Delete the Middle Node of a Linked List/IterativeSolution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        if(head==null || head.next==null)
            return null;

        ListNode prev = head, slow=head, fast=head;
        
        while(fast!=null && fast.next!=null) {
            prev=slow;
            slow=slow.next;
            fast=fast.next.next;
        }
        
        //prev is just before the middle node i.e. slow 
        prev.next= prev.next.next;
        
        return head;
    }
}

File: LinkedList Java and Kotlin/Delete the Middle Node of a Linked List/IterativeSolution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun deleteMiddle(head: ListNode?): ListNode? {
        if(head==null || head.next==null)
            return null;

        //We need to reduce one iteration 
        //thus, fast = fast.next.next
        // So that one slow.next statement is reduced
        // and slow points at element just before middle
        var slow:ListNode? = head
        var fast:ListNode? = head.next.next
        
        while(fast!=null && fast.next!=null) {
            slow=slow!!.next
            fast=fast!!.next?.next
        }
        
        slow?.next= slow?.next?.next
        
        return head
    }
}
